### PR TITLE
bitcoin: update to 25.1

### PR DESCRIPTION
**Pull request by fanquake**
###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS x.y
Xcode x.y / Command Line Tools x.y.z

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [ ] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

**Comment by aeiouaeiouaeiouaeiouaeiouaeiou** [comment]
@fanquake the [documentation states](https://github.com/bitcoin/bitcoin/blob/master/doc/build-unix.md) that a build with Berkeley DB <4.8 is required for compatibility with old wallets. However, adding [db46](https://ports.macports.org/port/db46/) to `depends_build`/`depends_lib` brings nothing and the build config reports `libdb_cxx headers missing`.

Builds on macOS <10.14 [are failing](https://ports.macports.org/port/bitcoin/builds/) because of this bug. Maybe we should block the search for these headers and pass the `--without-bdb` flag?

**Comment by fanquake** [comment]
> that a build with Berkeley DB <4.8 is required for compatibility with old wallets. However, adding db46 to

Legacy wallet support requires BDB = 4.8. That is the only supported version. Could you try with bdb48?

Modern wallets use SQLite3. Any recent version is ok.

> Builds on macOS <10.14 are failing because of this bug

The project no-longer supports this version of macOS, only 10.15 or later. So what to do in this case is up to Macports.

**Comment by fanquake** [comment]
I've rebased, and added `port:db48` to the wallet variant of the build.

**Comment by aeiouaeiouaeiouaeiouaeiouaeiou** [comment]
```
  Undefined symbols for architecture x86_64:
    "DbMpoolFile::get_fileid(unsigned char*)", referenced from:
        wallet::BerkeleyDatabase::Open() in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
    "DbMpoolFile::set_flags(unsigned int, int)", referenced from:
        wallet::BerkeleyDatabase::Open() in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
    "Db::remove(char const*, char const*, unsigned int)", referenced from:
        wallet::BerkeleyDatabase::Rewrite(char const*) in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
    "Db::rename(char const*, char const*, char const*, unsigned int)", referenced from:
        wallet::BerkeleyDatabase::Rewrite(char const*) in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
    "Db::verify(char const*, char const*, std::__1::basic_ostream<char, std::__1::char_traits<char> >*, unsigned int)", referenced from:
        wallet::BerkeleyDatabase::Verify(bilingual_str&) in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
    "Db::Db(DbEnv*, unsigned int)", referenced from:
        wallet::BerkeleyDatabase::Verify(bilingual_str&) in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
        wallet::BerkeleyDatabase::Open() in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
        wallet::BerkeleyDatabase::Rewrite(char const*) in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
    "Db::~Db()", referenced from:
        wallet::BerkeleyDatabase::Verify(bilingual_str&) in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
        wallet::BerkeleyDatabase::Rewrite(char const*) in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
    "Dbc::get(Dbt*, Dbt*, unsigned int)", referenced from:
        wallet::BerkeleyCursor::Next(DataStream&, DataStream&) in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
    "Dbc::close()", referenced from:
        wallet::BerkeleyCursor::~BerkeleyCursor() in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
    "Dbt::Dbt(void*, unsigned int)", referenced from:
        wallet::BerkeleyDatabase::Rewrite(char const*) in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
        wallet::BerkeleyBatch::ReadKey(DataStream&&, DataStream&) in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
        wallet::BerkeleyBatch::WriteKey(DataStream&&, DataStream&&, bool) in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
        wallet::BerkeleyBatch::EraseKey(DataStream&&) in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
        wallet::BerkeleyBatch::HasKey(DataStream&&) in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
    "Dbt::Dbt()", referenced from:
        wallet::BerkeleyCursor::Next(DataStream&, DataStream&) in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
        wallet::BerkeleyBatch::ReadKey(DataStream&&, DataStream&) in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
    "Dbt::~Dbt()", referenced from:
        wallet::SafeDbt::~SafeDbt() in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
        wallet::BerkeleyDatabase::Rewrite(char const*) in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
    "DbEnv::remove(char const*, unsigned int)", referenced from:
        wallet::BerkeleyEnvironment::Close() in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
    "DbEnv::version(int*, int*, int*)", referenced from:
        wallet::BerkeleyDatabase::Verify(bilingual_str&) in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
        wallet::BerkeleyDatabaseSanityCheck() in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
    "DbEnv::strerror(int)", referenced from:
        wallet::BerkeleyEnvironment::Close() in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
        wallet::BerkeleyEnvironment::Open(bilingual_str&) in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
    "DbEnv::DbEnv(unsigned int)", referenced from:
        wallet::BerkeleyEnvironment::Close() in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
        wallet::BerkeleyEnvironment::Reset() in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
    "DbEnv::~DbEnv()", referenced from:
        wallet::BerkeleyEnvironment::Close() in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
    "DbTxn::abort()", referenced from:
        wallet::BerkeleyBatch::Close() in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
        wallet::BerkeleyBatch::TxnAbort() in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
    "DbTxn::commit(unsigned int)", referenced from:
        wallet::BerkeleyBatch::TxnCommit() in libbitcoin_wallet.a(libbitcoin_wallet_a-bdb.o)
  ld: symbol(s) not found for architecture x86_64
  clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

GitHub Actions macOS-12 builder finished with this result, I had the same linker error on macOS 10.6 and 10.8 VMs.

**Comment by aeiouaeiouaeiouaeiouaeiouaeiou** [comment]
@jmroot @ryandesign @mascguy may need some help

File: finance/bitcoin/Portfile [hunk 1]
```diff
@@ -112,8 +112,12 @@ variant gui description {Build the Qt5 GUI} {
 }
 
 variant wallet description {Build with support for wallet} {
-    depends_lib-append port:sqlite3
+    depends_lib-append port:db48 \
+                       port:sqlite3
     configure.args-replace --enable-wallet=no --enable-wallet=yes
+    configure.env-append \
+        BDB_CFLAGS=-I${prefix}/include/db48 \
+        BDB_LIBS="-L${prefix}/lib/db48 -ldb_cxx-4.8"
```
Comment (jmroot) [inline finance/bitcoin/Portfile:120]: This is still quoted incorrectly, see my previous comment.

Comment (jmroot) [reply finance/bitcoin/Portfile:120]: This is the part in the log that tells you there was a problem with the env option, BTW:
```
DEBUG: parse_environment: skipping invalid entry: '-ldb_cxx-4.8"'
```

Comment (fanquake) [reply finance/bitcoin/Portfile:120]: Sorry, done this properly now.

**Comment by fanquake** [comment]
CI looking better now, thanks for the help here.

File: finance/bitcoin/Portfile [hunk 2]
```diff
@@ -12,7 +12,7 @@ legacysupport.use_mp_libcxx \
 
 name                    bitcoin
 categories              finance crypto
-version                 25.0
+version                 25.1
 revision                0
 platforms               darwin
```
Comment (herbygillot) [inline finance/bitcoin/Portfile:17]: `platforms darwin` is implicit, we can drop this line

Comment (fanquake) [reply finance/bitcoin/Portfile]: Dropped this line.

**Review by aeiouaeiouaeiouaeiouaeiouaeiou — APPROVED** [review]
LGTM

**Comment by herbygillot** [comment]
Thank you @fanquake for this submission, and thanks @aeiouaeiouaeiouaeiouaeiouaeiou & @jmroot for reviewing.

**Pull request merged**